onnxruntime/core/providers/cuda/tensor/gather_elements_impl.cu

```diff
@@ -16,74 +16,84 @@ namespace cuda {
 namespace {
 constexpr int kThreadsPerBlock = GridDim::maxThreadsPerBlock;
 constexpr int kThreadWorkSize = 4;
-}  // namespace
-
-template <class T>
-struct FuncAssignment {
-  __device__ __inline__ void operator()(T* a, const T* b) const { *a = *b; }
-};
 
-template <typename T, typename TIndex, bool IsOuterAxis, bool IsGather, typename TFunc>
-__global__ void _GatherScatterElements2DKernel(const T* src_data, const TIndex* indices_data, T* output_data,
-                                               const int64_t input_dim_along_axis,
-                                               const fast_divmod indices_row_size_fmd, const int64_t input_row_size,
-                                               const TFunc& func, CUDA_LONG N) {
-  CUDA_LONG start = kThreadsPerBlock * kThreadWorkSize * blockIdx.x + threadIdx.x;
-  CUDA_LONG id;
-  T value[kThreadWorkSize];
+template <bool IsStridedIndices>
+struct OffsetCalculator {
+  OffsetCalculator(const int rank, const TArray<int64_t> masked_input_strides, const TArray<fast_divmod> indices_fdms,
+                   const TArray<int64_t> indices_strides)
+      : rank_(rank), indices_fdms_(indices_fdms) {
+    masked_input_strides_.SetSize(rank);
+    if (IsStridedIndices) indices_strides_.SetSize(rank);
+    for (int dim = 0; dim < rank; ++dim) {
+      masked_input_strides_[dim] = static_cast<CUDA_LONG>(masked_input_strides[dim]);
+      if (IsStridedIndices) indices_strides_[dim] = static_cast<CUDA_LONG>(indices_strides[dim]);
+    }
+  }
 
-  if (!IsGather) {
-    id = start;
+  __device__ __forceinline__ TArray<CUDA_LONG, 2> get(CUDA_LONG linear_idx) const {
```

Comment:
could you please add comment explaining the function, especially the meaning of the elements in the outputs.

Reply:
Added. But to me code is much more meaningful than comments, and sometimes comments confuse me a lot than code. :-)